src-self-hosted/zir_sema.zig

```diff
@@ -114,15 +113,10 @@ pub fn analyzeBody(mod: *Module, scope: *Scope, body: zir.Module.Body) !void {
 
 pub fn analyzeBodyValueAsType(mod: *Module, block_scope: *Scope.Block, body: zir.Module.Body) !Type {
     try analyzeBody(mod, &block_scope.base, body);
-    for (block_scope.instructions.items) |inst| {
-        if (inst.castTag(.ret)) |ret| {
-            const val = try mod.resolveConstValue(&block_scope.base, ret.operand);
-            return val.toType();
-        } else {
-            return mod.fail(&block_scope.base, inst.src, "unable to resolve comptime value", .{});
-        }
-    }
-    unreachable;
+
+    const result_type = body.instructions[body.instructions.len - 1];
+    const val = try mod.resolveConstValue(&block_scope.base, result_type.analyzed_inst.?);
+    return val.toType();
```

Comment:
The "set return value" instruction would be less brittle than looking at the last instruction too, I think.